### PR TITLE
Wrap jackson optimization modules with a factory API

### PR DESCRIPTION
**Pull request by carterkozak**
This allows us greater control, more quickly, over jackson optimization
modules.

==COMMIT_MSG==
Wrap jackson optimization modules with a factory API
==COMMIT_MSG==

This module is pretty trivial, there's not much to test. Maybe the boolean logic for `NO_OPTIMIZATIONS`, however static init is exceedingly hard to test, and the structure built to enable graalvm prune unreachable code.

### CODE REVIEW

File: conjure-java-jackson-optimization/src/main/java/com/palantir/conjure/java/optimizations/ObjectMapperOptimizations.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ * This allows us to quickly, transparently change between implementations such as Afterburner, Blackbird, or none at
+ * all.
+ */
+public final class ObjectMapperOptimizations {
+
+    private static final boolean NO_OPTIMIZATIONS =
+            // These optimizations are not used within graalvm because they leverage dynamic class creation
+            System.getProperty("org.graalvm.nativeimage.imagecode") != null
+                    // This may be globally disabled with a system property
+                    || "true".equalsIgnoreCase(System.getProperty("com.palantir.conjure.java.optimizations.disabled"));
```
Comment (jkozlowski) [inline conjure-java-jackson-optimization/src/main/java/com/palantir/conjure/java/optimizations/ObjectMapperOptimizations.java]: Is optimizations too generic? What if you have some other "optimizations" in conjure that you'd like to disable?

Comment (carterkozak) [reply conjure-java-jackson-optimization/src/main/java/com/palantir/conjure/java/optimizations/ObjectMapperOptimizations.java]: Would those be optimizations for the ObjectMapper itself, or more specific?

Comment (jkozlowski) [reply conjure-java-jackson-optimization/src/main/java/com/palantir/conjure/java/optimizations/ObjectMapperOptimizations.java]: Basically all I'm suggesting is "com.palantir.conjure.java.optimizations.disabled" -> "com.palantir.conjure.java.jackson.optimizations.disabled", but mostly a nit

**Review by jkozlowski — APPROVED** [review]
Just a comment

File: conjure-java-jackson-optimization/src/main/java/com/palantir/conjure/java/optimizations/ObjectMapperOptimizations.java (hunk 1 above)
Comment (carterkozak) [reply conjure-java-jackson-optimization/src/main/java/com/palantir/conjure/java/optimizations/ObjectMapperOptimizations.java]: Oh sorry, I misinterpreted your comment! Updating the package name and disable string totally make sense 👍

Comment (carterkozak) [reply conjure-java-jackson-optimization/src/main/java/com/palantir/conjure/java/optimizations/ObjectMapperOptimizations.java]: Not sure how I got it in my head that the comment referred to the class name `ObjectMapperOptimizations`. Updated!

**Pull request merged**